bottlecap/src/appsec/processor/context.rs

```diff
@@ -24,25 +30,134 @@ pub struct Context {
     waf: WafContext,
     /// The timeout for the WAF.
     waf_timeout: Duration,
+    /// The total time spent on WAF runs.
+    waf_duration: Duration,
+    /// The number of times the WAF timed out during this request.
+    waf_timed_out_occurrences: usize,
     // Whether this context has events.
     has_events: bool,
     /// Trace tags to be applied for this invocation.
     trace_tags: HashMap<TagName, TagValue>,
+    /// Whether we have had a chance to see the response body or not.
+    response_seen: bool,
+    /// Holds a trace for sending once the response is seen.
+    on_response_seen: Option<(Vec<Span>, SendingTraceProcessor, HoldArguments)>,
 }
 impl Context {
     /// Creates a new [`Context`] for the provided request ID and configures it
     /// with the provided WAF timeout.
-    pub(crate) fn new(rid: String, waf_handle: &mut Handle, waf_timeout: Duration) -> Self {
+    pub(crate) fn new(
+        rid: String,
+        waf_handle: &mut Handle,
+        ruleset_version: &str,
+        waf_timeout: Duration,
+    ) -> Self {
+        let mut trace_tags = HashMap::from([
+            (TagName::AppsecEnabled, TagValue::Metric(1.0)),
+            (
+                TagName::AppsecWafVersion,
+                TagValue::Always(libddwaf::get_version().to_string_lossy().to_string()),
+            ),
+        ]);
+        if !ruleset_version.is_empty() {
+            trace_tags.insert(
+                TagName::AppsecEventRulesVersion,
+                TagValue::Always(ruleset_version.to_string()),
+            );
+        }
+
         Self {
             rid,
             expected_response_format: ExpectedResponseFormat::default(),
             waf: waf_handle.new_context(),
             waf_timeout,
+            waf_duration: Duration::ZERO,
+            waf_timed_out_occurrences: 0,
             has_events: false,
-            trace_tags: HashMap::from([(TagName::AppsecEnabled, TagValue::Metric(1.0))]),
+            trace_tags,
+            response_seen: false,
+            on_response_seen: None,
```

Comment:
Oh yeah having `response_seen` and `on_response_seen` confused me a lot, I just saw these are two different things

Reply:
It's a "classic" `finalized`/`onFinalize` pair. But we can model this a little more to make it more obvious.